godot-core/src/builtin/dictionary.rs

```diff
@@ -0,0 +1,343 @@
+        }
+        dict
+    }
+}
+
+/// Creates a `Dictionary` from the given `HashSet`. Each key is converted
+/// to a `Variant`, and the values are all `true`.
+impl<K: ToVariant> From<&HashSet<K>> for Dictionary {
+    fn from(set: &HashSet<K>) -> Self {
+        let mut dict = Dictionary::new();
+        for k in set.into_iter() {
+            dict.insert(k.to_variant(), true.to_variant());
+        }
+        dict
+    }
+}
+
+/// Inserts all key-values from the iterator into the dictionary,
+/// replacing values with existing keys with new values returned
+/// from the iterator.
+impl<K: ToVariant, V: ToVariant> Extend<(K, V)> for Dictionary {
+    fn extend<T: IntoIterator<Item = (K, V)>>(&mut self, iter: T) {
+        for (k, v) in iter.into_iter() {
+            self.insert(k.to_variant(), v.to_variant());
+        }
+    }
+}
+
+impl<K: ToVariant, V: ToVariant> FromIterator<(K, V)> for Dictionary {
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
+        let mut dict = Dictionary::new();
+        for (k, v) in iter.into_iter() {
+            dict.insert(k.to_variant(), v.to_variant());
+        }
+        dict
+    }
+}
+
+impl_builtin_traits! {
+    for Dictionary {
+        Default => dictionary_construct_default;
+        Drop => dictionary_destroy;
+        Eq => dictionary_operator_equal;
+    }
+}
+
+impl GodotFfi for Dictionary {
+    ffi_methods! { type sys::GDExtensionTypePtr = *mut Opaque; .. }
+}
+
+impl fmt::Debug for Dictionary {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.to_variant().stringify())
+    }
+}
+
+/// Creates a new reference to the data in this dictionary. Changes to the original dictionary will be
+/// reflected in the copy and vice versa.
+///
+/// To create a (mostly) independent copy instead, see [`Dictionary::duplicate_shallow()`] and
+/// [`Dictionary::duplicate_deep()`].
+impl Share for Dictionary {
+    fn share(&self) -> Self {
+        unsafe {
+            Self::from_sys_init(|self_ptr| {
+                let ctor = sys::builtin_fn!(dictionary_construct_copy);
+                let args = [self.sys_const()];
+                ctor(self_ptr, args.as_ptr());
+            })
+        }
+    }
+}
+
+/// Creates a new dictionary with the given keys and values, the syntax mirrors
+/// godot's dictionary creation syntax.
+///
+/// Currently only literal keys are supported. But any expression whose result
+/// can be converted with `Variant::from()` may be used as a value.
+///
+/// Example
+/// ```rust, no_run
+/// # #[macro_use] extern crate godot_core;
+/// # fn main() {
+/// let d = dict! {
+///     "key1": 10,
+///     "another": Variant::nil(),
+///     "true": true,
+///     "final": "final",
+/// }
+/// # }
+/// ```
+#[macro_export]
+macro_rules! dict {
+    ($($key:literal: $value:expr),* $(,)?) => {
+        {
+            let mut d = ::godot::builtin::Dictionary::new();
```

Comment:
Maybe use `$crate` instead of `::godot`? I'm always confused about macro scoping. Consider using it in your integration tests to make sure it's actually usable.